conda_smithy/configure_feedstock.py

```diff
@@ -391,6 +391,16 @@ def _collapse_subpackage_variants(
 
     for k, v in squished_used_variants.items():
         if k in all_used_vars:
+            input_variant = squished_input_variants.get(k)
+            if input_variant and isinstance(input_variant, (tuple, list)):
+                # Ensure we retain the input order to avoid mismatched entries
+                # with the preserve_top_level_loops overrides below.
+                # NOTE: "pin_run_as_build" (dict of dict) order is not handled.
+                # "tuple" below only used for uniform list/tuple/str comparison.
+                v_set = set(map(tuple, v))
+                v = type(input_variant)(
+                    v_i for v_i in input_variant if tuple(v_i) in v_set
+                )
```

Comment:
I opted to no handle `pin_run_as_build` since its order *should* not matter, i.e., we'd had to do special handling for that case here with no further benefit.
I added at least a note so we are aware that it's intentionally omitted.

The variant entries should be of type `str`, so a straight `v_set = set(v)` would be possible.
We do, however, also have `zip_keys` which is a `list` of `list`/`tuple`.
To handle potential `tuple`-to-`list` comparisons (and to get hashable types for `list` entries) this just casts everything to `tuple`s.

Feel free to amend any of this in case you'd want more explicit handling.